feat(markdown): add `excludeLangs` configuration to syntax highlighting (#13311)
withastro/astro: .changeset/large-dolphins-roll.md

---
'@astrojs/markdown-remark': minor
'astro': minor
---

Adds support for `excludeLangs` in `syntaxHighlight`. This allows users to exclude languages from syntax highlighting. This is useful when rendering markdown that contains code blocks with specific languages which has integrations that depend on the language not being highlighted, such as mermaid diagrams.

> The default value for `excludeLangs` is `['math']` and is unchanged in this release.

Reviewer: Noting that I will wait to review the changeset after we have docs, but if something is new, then how can its default value be unchanged? (compared to what?)

I'm confused by what this is trying to communicate, but don't want to miss something important! 😄

Author: Updated this.

The following example configuration will exclude highlighting for `mermaid` and `math` code blocks:

```js
import { defineConfig } from 'astro/config';

export default defineConfig({
  markdown: {
    syntaxHighlight: {
      type: 'shiki',

>       excludeLangs: ['mermaid', 'math'],

Author: Should this be `excludeLanguages`? 
Is there a convention for astro configurations?

---
'@astrojs/markdown-remark': minor
'astro': minor
---

> Adds support for `excludeLangs` to allow users to customize syntax highlighting behavior to exclude languages from syntax highlighting. This is useful when rendering markdown that contains code blocks with specific languages with integrations that depend on the code not being highlighted, such as mermaid diagrams. Mermaid diagrams are rendered using the code block with the language `mermaid`. Syntax highlighting introduces additional code blocks with the language `astro` and `html` to wrap the diagram, which is redundant when the diagram is rendered using a different integration.

Reviewer: ```suggestion
Adds a new configuration option for Markdown syntax highlighting `excludeLangs` 

This option provides better support for diagramming tools that rely on Markdown code blocks, such as Mermaid.js and D2 by allowing you to exclude specific languages from Astro's default syntax highlighting.

This option allows you to avoid rendering conflicts with tools that depend on the code not being highlighted without forcing you to disable syntax highlighting for other code blocks.
```
Just tightened this up a bit, and took from the blog post!

---
'@astrojs/markdown-remark': minor
'astro': minor
---

Adds a new configuration option for Markdown syntax highlighting `excludeLangs` 

This option provides better support for diagramming tools that rely on Markdown code blocks, such as Mermaid.js and D2 by allowing you to exclude specific languages from Astro's default syntax highlighting.

This option allows you to avoid rendering conflicts with tools that depend on the code not being highlighted without forcing you to disable syntax highlighting for other code blocks.

> The default value for `excludeLangs` is `['math']` and remains unchanged by default in this release.
> But users can now override it to exclude other languages or exclude no languages.

Reviewer: ```suggestion
```
I don't think this is necessary here. This doesn't have to be docs, but more an introduction to let people know about the new feature. Adding too many details can detract from just giving people the main idea so they can decide if this is something they care about/want to read more about.

The following example configuration will exclude highlighting for `mermaid` and `math` code blocks:

```js
import { defineConfig } from 'astro/config';

export default defineConfig({
  markdown: {
    syntaxHighlight: {
      type: 'shiki',
      excludeLangs: ['mermaid', 'math'],
    },
  },
});
```

> 

Reviewer: ```suggestion

Read more about this new option in the [Markdown syntax highlighting configuration docs](https://docs.astro.build/en/reference/configuration-reference/#markdownsyntaxhighlight).
```
